contracts/button-token/lib.rs

```diff
@@ -0,0 +1,572 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract]
+mod button_token {
+    use ink_storage::{traits::SpreadAllocate, Mapping};
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct ButtonToken {
+        /// Total token supply.
+        total_supply: Balance,
+        /// Mapping from owner to number of owned token.
+        balances: Mapping<AccountId, Balance>,
+        /// Mapping of the token amount which an account is allowed to withdraw
+        /// from another account.
+        allowances: Mapping<(AccountId, AccountId), Balance>,
+        /// access control
+        owner: AccountId,
+    }
+
+    /// Event emitted when a token transfer occurs.
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct Transfer {
+        #[ink(topic)]
+        pub from: Option<AccountId>,
+        #[ink(topic)]
+        pub to: Option<AccountId>,
+        pub value: Balance,
+    }
+
+    /// Event emitted when an approval occurs that `spender` is allowed to withdraw
+    /// up to the amount of `value` tokens from `owner`.
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct Approval {
+        #[ink(topic)]
+        owner: AccountId,
+        #[ink(topic)]
+        spender: AccountId,
+        value: Balance,
+    }
+
+    /// Event emitted when TheButton owner is changed
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct OwnershipTransferred {
+        #[ink(topic)]
+        from: AccountId,
+        #[ink(topic)]
+        to: AccountId,
+    }
+
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        /// Returned if not enough balance to fulfill a request is available.
+        InsufficientBalance,
+        /// Returned if not enough allowance to fulfill a request is available.
+        InsufficientAllowance,
+        /// Returned when an account which is not the owner calls a method with access control
+        NotOwner,
+    }
+
+    /// Result type    
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    impl ButtonToken {
+        /// Creates a new contract with the specified initial supply.
+        #[ink(constructor)]
+        pub fn new(initial_supply: Balance) -> Self {
+            // This call is required in order to correctly initialize the
+            // `Mapping`s of our contract.
+            ink_lang::utils::initialize_contract(|contract| {
+                Self::new_init(contract, initial_supply)
+            })
+        }
+
+        /// Default initializes the contract with the specified initial supply.
+        fn new_init(&mut self, initial_supply: Balance) {
+            let caller = Self::env().caller();
```

Comment:
Short answer: definitely.

Long answer: yes, but how?
- a baked in list of addresses? Ugly and problematic, error prone.
- another contract which acts as an access controller to this one?
- if we have access from the contract level  to the address that deployed the code, this address itself could be made the default whitelisted can_initialize ?

Reply:
I can imagine a pallet, like 'contract-owner', which effectively keeps a map: `String -> (AccountId, Vec<AccountId)`, where entry `(name -> (author, owners))` means that:
 - `author` has reserved key `name` and thus only `author` can modify `owners` vec (at any time)
 - in the contract code you can guard your constructor for accounts specified in `contract-owner` pallet's map under some key

but this seems too much complicated and makes testing much harder. I will keep this discussion open, so that other reviewers may find some solution

Reply:
I asked on SO, and there is no way to get the code owner from the contract env level:
https://substrate.stackexchange.com/questions/3181/how-to-get-contracts-code-owner/3184#3184

This is so useful, given this SC model, that we might want to think about creating a functionality that allows contracts to query the runtime for their code owner's. There is an [example](https://github.com/paritytech/ink/tree/master/examples/rand-extension) for creating runtime-contract call.

Reply:
afaik, when writing `ChainExtension` you have to be extremely careful, I hope there is another way

Reply:
WYM by _extremely careful_? YMMV but I think this is a relatively safe, specific functionality - and in this case it will be read-only anyway?

Reply:
quoting from Substrate:
> It cannot be overstated how delicate of a process the creation of a chain extension is. Check whether using [`bare_call`](crate::Pallet::bare_call) suffices for the use case at hand.